ray_provider/decorators/ray_decorators.py

```diff
@@ -64,9 +73,111 @@ def sum_cols(df: pd.DataFrame) -> pd.DataFrame:
     @functools.wraps(python_callable)
     def wrapper(f):
 
+        def on_execute_callback(context):
+            pass
+
+        @provide_session
```

Comment:
<A> -> <B> -> <C>
Say A succeeeds, B fails

then upon retry of B, A needs to have its output downloaded, B should continue as if it never ran.